packages/lodestar/src/network/encoders/response.ts

```diff
@@ -0,0 +1,107 @@
+      let compressor = getCompressor(encoding);
+      if(!type) {
+        return;
+      }
+      for await (const chunk of source) {
+        if(chunk.status !== 0) {
+          yield encodeResponseStatus(chunk.status);
+          break;
+        }
+        //yield status
+        yield encodeResponseStatus(chunk.status);
+        let serializedData: Buffer;
+        try {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const serialized = type.serialize(chunk.body as any);
+          serializedData = Buffer.from(serialized.buffer, serialized.byteOffset, serialized.length);
+        } catch (e) {
+          logger.warn(`Failed to ssz serialize chunk of method ${method}. Error: ${e.message}`); 
+        }
+        //yield encoded ssz length
+        yield Buffer.from(encode(serializedData.length));
+        //yield compressed or uncompressed data chunks
+        yield* compressor(serializedData);
+        //reset compressor
+        compressor = getCompressor(encoding);
+      }
+    })();
+  });
+}
+
+export function eth2ResponseDecode(
+  config: IBeaconConfig, logger: ILogger, method: Method, encoding: ReqRespEncoding
+): (source: AsyncIterable<Buffer>) => AsyncGenerator<ResponseBody> {
+  return (source) => {
+    return (async function*() {
+      //floating buffer with current chunk
+      let buffer = new BufferList();
+      //holds uncompressed chunks
+      let uncompressedData = new BufferList();
+      let status: number = null;
+      let sszLength: number = null;
+      const decompressor = getDecompressor(encoding);
+      const type = Methods[method].responseSSZType(config);
+      for await (const chunk of source) {
+        buffer.append(chunk);
+        if(status === null) {
+          status = buffer.get(0);
+          buffer.consume(1);
+          if(status !== RpcResponseStatus.SUCCESS) {
+            logger.warn(`Received err status ${status} for method ${method}`);
+            break;
+          }   
+        }
+        if(buffer.length === 0) continue;
+        if(sszLength === null) {
+          sszLength = decode(buffer.slice());
+          buffer.consume(decode.bytes);
+        }
+        if(buffer.length === 0) continue;
+        let uncompressed: Buffer;
+        try {
+          uncompressed = decompressor.uncompress(buffer.slice());
+          buffer.consume(buffer.length);
+        } catch (e) {
+          logger.warn(`Failed to uncompress data for method ${method}. Error: ${e.message}`);
+        }
+        if(uncompressed) {
+          uncompressedData.append(uncompressed);
+          if(uncompressedData.length > sszLength) {
+            logger.warn(`Received too much data for method ${method}`);
+            break;
+          }
+          if(uncompressedData.length === sszLength) {
+            try {
+              yield type.deserialize(uncompressedData.slice()) as ResponseBody;
+              buffer = new BufferList();
```

Comment:
what's the point to reuse `buffer` and every variable except for `decompressor ` instead of just declare them inside the `for` loop?

Reply:
I could call consume on buffer and uncompressData to remove all buffers but I feel like this is faster as it will leave current data for gc to clean while consume might be fiddling with existing arrays. Decompressor has reset method to this cleanup. Once we deserialize data it means we reached end of chunk, and next one will be exactly the same so we just reset everything. I  should add more code comments here

Reply:
@mpetrunic yes please! Comments would be very helpful since there are many non-obvious design choices that are really good and should be preserved if the code is re-worked